lib/SILOptimizer/Mandatory/Differentiation.cpp

```diff
@@ -3397,45 +3810,114 @@ class AdjointEmitter final : public SILInstructionVisitor<AdjointEmitter> {
       return;
     }
     auto applyInfo = applyInfoLookUp->getSecond();
-    auto origTy = ai->getCallee()->getType().castTo<SILFunctionType>();
-    SILFunctionConventions origConvs(origTy, getModule());
 
     // Get the pullback.
     auto *field = getPrimalInfo().lookUpPullbackDecl(ai);
     assert(field);
     auto loc = ai->getLoc();
-    SILValue pullback = builder.createStructExtract(loc,
-                                                    primalValueAggregateInAdj,
-                                                    field);
+    SILValue pullback =
+        builder.createStructExtract(loc, primalValueAggregateInAdj, field);
 
-    // Construct the pullback arguments.
+    // Get the original result of the `apply` instruction.
     SmallVector<SILValue, 8> args;
-    auto av = takeAdjointValue(ai);
-    ValueWithCleanup seedBuf(builder.createAllocStack(loc, av.getType()),
+    SmallVector<SILValue, 8> origDirResults;
+    collectAllExtractedElements(ai, origDirResults);
+    SmallVector<SILValue, 8> origAllResults;
+    collectAllActualResultsInTypeOrder(
+        ai, origDirResults, ai->getIndirectSILResults(),
+        origAllResults);
+    auto origResult = origAllResults[applyInfo.actualIndices.source];
+    auto origNumIndRes = ai->getNumIndirectResults();
+
+    auto pullbackType =
+        remapType(pullback->getType()).castTo<SILFunctionType>();
+    SILFunctionConventions pullbackConvs(pullbackType, getModule());
+
+    // Get adjoint value of the original result.
+    SILType adjointType;
+    SILValue adjointValue;
+    if (origResult->getType().isObject()) {
+      // If original result is a `tuple_extract`, materialize adjoint value of
+      // `ai` and extract the corresponding element adjoint value.
+      if (auto *tupleExtract = dyn_cast<TupleExtractInst>(origResult)) {
+        auto adjointTuple = materializeAdjoint(takeAdjointValue(ai), loc);
+        adjointValue = builder.emitTupleExtract(loc, adjointTuple,
+                                                tupleExtract->getFieldNo());
+      }
+      // Otherwise, materialize adjoint value of `ai`.
+      else {
+        adjointValue = materializeAdjoint(takeAdjointValue(origResult), loc);
+      }
+      adjointType = adjointValue->getType();
+    } else {
+      auto adjointBuf = getAdjointBuffer(origResult);
+      adjointValue = adjointBuf.getValue();
+      adjointType = adjointBuf.getValue()->getType();
+    }
+
+    // Register indirect results as arguments.
```

Comment:
Can you elaborate a bit more on what this is doing? I don't understand what `tempAllocation` is for.

Reply:
Wow, the original code was completely wrong.

Things are fixed and made clear in https://github.com/apple/swift/pull/22544/commits/6f068338919019d8ae7d1178900d5ada0d050c7d.
The code here in `AdjointEmitter::visitApplyInst` allocates buffers for pullback indirect results. Those indirect result buffers may then be accumulated into original indirect parameters' adjoint buffers, depending on the differentiation parameter indices.